### ISSUE TITLE
server: /embeddings endpoint could/should accept list[int] as well as list[str]

### ISSUE THREAD

**Issue by gkluoe** [abetlen/llama-cpp-python#547]
I'm not sure whether this is a pedantic 1 -> 2 issue, or whether the underlying llama.cpp just doesn't support input for embedding in numerical tokenised form. I should note at this point that I'm very much a newbie in this, and if I completely have the wrong end of the stick please gently tell me so!

I am using the langchain's `OpenAIEmbedding` to try to talk to llama-cpp-python's API server to retrieve embeddings. It appears that langchain converts the input into tokens in the form of a list of `int` before calling the /v1/embeddings endpoint with that list. 

Llama server doesn't like this:

`Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIError: [{'type': 'string_type', 'loc': ('body', 'input', 'str'), 'msg': 'Input should be a valid string', 'input': [[8586, 374, 1063, 1495]], 'url': 'https://errors.pydantic.dev/2.1/v/string_type'}, {'type': 'string_type', 'loc': ('body', 'input', 'list[str]', 0), 'msg': 'Input should be a valid string'`

OpenAI's API will accept input to the `embeddings` endpoint as a string, a list of strings, or a list of `int` so I presume it would be desirable for this server to do the same.

I suspect that when using pydantic 1, `int` would be coerced to strings transparently, so no error would be thrown at this stage; however in v2:

> Validator behavior changes (https://docs.pydantic.dev/latest/migration/#validator-behavior-changes)
        Pydantic V2 includes some changes to type coercion. For example:
             int, float, and decimal values are no longer coerced to strings.

# Expected Behavior

The OpenAI `embeddings` API documentation states it accepts 'string or array of tokens' and it does indeed appear to accept an array of tokens in bytes or `int` form.

# Current Behavior

Calling /v1/embeddings with an array of `int` (see reproducible command below) produces the following error:

`Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIError: [{'type': 'string_type', 'loc': ('body', 'input', 'str'), 'msg': 'Input should be a valid string', 'input': [[8586, 374, 1063, 1495]], 'url': 'https://errors.pydantic.dev/2.1/v/string_type'}, {'type': 'string_type', 'loc': ('body', 'input', 'list[str]', 0), 'msg': 'Input should be a valid string'`

# Environment and Context

* Physical (or virtual) hardware you are using:

(Nb I am running in docker - this is the host hardware)

  Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 57 bits virtual
  Byte Order:            Little Endian
  CPU(s):                  8
  On-line CPU(s) list:   0-7

* Operating System

Linux 5.15.0-78-generic #85-Ubuntu SMP Fri Jul 7 15:25:09 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

* SDK version

```
Python 3.10.6 (in container)
GNU Make 4.3
gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
```

# Steps to Reproduce

Call the embeddings API with a list of int rather than strings:

This works OK:

```
curl -X 'POST' \
  'http://localhost/v1/embeddings' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "input": "Hello world!"
}'
```

This also works OK (though presumably provides embeddings for tokens that look like numbers rather than the numerical values of the tokens?)
```
curl -X 'POST' \
  'http://localhost/v1/embeddings' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "input": ["8586", "374", "1063", "1495"]
}'
```

This fails:
```
curl -X 'POST' \
  'http://localhost/v1/embeddings' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "input": [8586, 374, 1063, 1495]
}'
```

But the same operation against the OpenAPI succeeds:
```
curl https://api.openai.com/v1/embeddings \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer sk-xxxxxxxxxxxxx" \
  -d '{
    "input": [8586, 374, 1063, 1495]
  }'


# Failure Logs

```
Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIError: [{'type': 'string_type', 'loc': ('body', 'input', 'str'), 'msg': 'Input should be a valid string', 'input': [[8586, 374, 1063, 1495]], 'url': 'https://errors.pydantic.dev/2.1/v/string_type'}, {'type': 'string_type', 'loc': ('body', 'input', 'list[str]', 0), 'msg': 'Input should be a valid string'
```

environment info:
```
~/git/llama-cpp-python$ git log | head -1
commit d89bf8c0722e65ad1a46a8ac4495dbf57ddfcb18

llama-cpp-python$ pip list | egrep "uvicorn|fastapi|sse-starlette|numpy|pydantic"
fastapi           0.100.1
numpy             1.25.1
pydantic          2.1.1
pydantic_core     2.4.0
pydantic-settings 2.0.2
sse-starlette     1.6.1
uvicorn           0.23.2
```

**Comment by c0sogi** [comment]
How about this handy implementation? This ensures that input value is list of strings.

```python
@router.post(
    "/v1/embeddings",
)
async def create_embedding(
    request: CreateEmbeddingRequest, llama: llama_cpp.Llama = Depends(get_llama)
):
    # Force input to be a list of int(s) / str(s)
    list_input: List = (
        request.input if isinstance(request.input, list) else [request.input]
    )

    # Force input to be a list of str
    request.input = (
        [  # Detokenize ids to strings, if they are integers
            llama.detokenize([id for id in list_input if isinstance(id, int)]).decode(
                "utf-8", errors="ignore"
            )
        ]
        if isinstance(list_input[0], int)
        else list_input  # This is a list of strings
    )  # type: ignore
    return await run_in_threadpool(
        llama.create_embedding, **request.model_dump(exclude={"user"})
    )
```

**Comment by jasoncaojingren** [comment]
Any update about this issue? I run latest version of llama-cpp-python 0.1.77, still got this problem when try to call it from LangChain. Because LangChain by default use the length-safe embedding approach.

    # please refer to
    # https://github.com/openai/openai-cookbook/blob/main/examples/Embedding_long_inputs.ipynb
    def _get_len_safe_embeddings(


Thanks for any update.

**Comment by gkluoe** [comment]
> However, I think OpenAIEmbedding uses tiktoken tokenizer, which is mostly like to be cl100k_base, which is not compatible with llama tokenizer. So, we need to make those openai-ids to be llama-ids forcibly somehow.

Good point. I suppose the simplest way would be to import tiktoken and use it to de-tokenise, but that would add a dependency.

**Comment by kicocisco** [comment]
@gkluoe Did you happen to find a work around for this?

**Comment by gkluoe** [comment]
I'm afraid not - as above my only real idea would be to insert tiktoken into the workflow to detokenise and the re-tokenise with llama tokeniser. I'm far from an external in this stuff though, nor do I have any time to actually try to make it work, sorry!

**Comment by kicocisco** [comment]
> 

Hey no worries, its a great idea and hopefully I find the time to implement. :)

Best!

**Comment by gkluoe** [comment]
Great - now that you've reminded me I will also try to find time to look at it again :)

**Comment by c0sogi** [comment]
Well, this is my logic to accomplish this in my project.

https://github.com/c0sogi/llama-api/blob/master/llama_api/logits/bias.py

I hope this might help!

**Comment by drdh** [comment]
A possible short-term alternative is to directly modify Langchain or create a new class `OpenAIEmbeddings`.

`input=encoding.decode(tokens[i : I + _chunk_size][0])`

[langchain/embeddings/openai.py#L376C39-L376C39](https://github.com/langchain-ai/langchain/blob/8e88ba16a88affc272da4bb99bd515b714781179/libs/langchain/langchain/embeddings/openai.py#L376C39-L376C39)
